Update with latest changes from IBM/simpool

davidbeckingsale opened the pull request.

jdahm committed:
Add License, change README.md

mcfadden8 committed:
splitBlock alignment fix from @mcfadden8

Fixed bug where we were not allocating to an aligned boundary everywhere even though splitBlock() was assuming we were.  This caused much badness

jdahm committed:
Rename PoolAllocator -> SizePool

jdahm committed:
Rename config file

jdahm committed:
Improve FixedSizePool and rename test files

jdahm committed:
Revert "splitBlock alignment fix from @mcfadden8"

This reverts commit 8cfa9942ae626a25e3df57031924374f1ee77182.

Causes the size tracking to break and thus the unit tests to fail.

jdahm committed:
Merge branch 'fixedsizealloc'

davidbeckingsale committed:
Merge remote-tracking branch 'upstream/master' into feature/merge-upstream

mcfadden8 commented on AllocatorTest.hpp:
```diff
@@ -81,7 +81,7 @@ struct STLAllocator {
   typedef std::size_t size_type;
 
-  typedef DynamicPoolAllocator<AllocatorType> PoolType;
+  typedef DynamicSizePool<AllocatorType> PoolType;
```
We will need to make sure and change all of our using code to use this new name.

mcfadden8 commented on DynamicSizePool.hpp:
```diff
@@ -24,8 +27,8 @@ class DynamicPoolAllocator
   // Allocator for the underlying data
-  typedef FixedPoolAllocator<struct Block, IA, (1<<6)> BlockAlloc;
-  BlockAlloc blockAllocator;
+  typedef FixedSizePool<struct Block, IA, IA, (1<<6)> BlockPool;
```
I'm not sure I understand this. Shouldn't this be "MA, IA" instead of "IA, IA"?

davidbeckingsale replied:
The new FixedPool can use different allocators for the meta-data and the actual data (e.g. for storing data in GPU memory). In this case, we want to use the same allocator for both. So I think this is correct.

mcfadden8 replied:
I'm not sure I see why FixedPool has use cases for separate allocators of meta-data and actual data and DynamicSizePool does not?  What is the unique use case(s) that FixedPool has that DynamicSizePool does not with respect to this?

davidbeckingsale replied:
FixedPool was storing the metadata and actual data all in the same allocation. DynamicPool does not (they are already separate).

mcfadden8 replied:
Okay.  I see.

davidbeckingsale committed:
Fixed missed merge bits

jdahm commented:
Let me know if I can help with this by changing anything in simpool itself that doesn't look right. I know I need to add better testing still, and I do want to incorporate @mcfadden8 point about block splitting. That's all on the todo list.

davidbeckingsale committed:
Add FixedSize pool that uses AllocationStrategy

davidbeckingsale commented:
@mcfadden8 - this should be good to go. Note that I have added a new pool that is basically a copy of the fixed-size pool, modified to use an AllocationStrategy. This allows us to still use the old one without an AllocationStrategy (faster) inside the dynamic pool.

davidbeckingsale committed:
Remove extra fixed pool

davidbeckingsale committed:
Correct merge and update CMakeLists

davidbeckingsale commented:
@mcfadden8 - this is ready and tested.

mcfadden8 commented on Makefile:
```diff
@@ -1,8 +1,8 @@
-# config.h can overwrite these defaults
 CXX ?= clang++
-CXXFLAGS ?= -O3
+CXXFLAGS ?= -O1 -g
```
Why did this change from -O3 to -O1?

davidbeckingsale replied:
No idea, but it doesn't impact us.

mcfadden8 replied:
Good point @davidbeckingsale, I should have noticed that this was in a Makefile.

jdahm commented:
Probably because I was testing something and forgot to set it back. I'll make that change now. EDIT: pushed.

mcfadden8 reviewed (approved):
Looks good

mcfadden8 commented:
@davidbeckingsale, can this branch be removed now that it has been merged?

davidbeckingsale commented:
Sure!
